src/charm.py

```diff
@@ -126,43 +131,50 @@ def _on_cluster_relation_changed(self, event: EventBase) -> None:
             return
 
         # check whether restart is needed for all `*_changed` events
-        self.on[f"{self.restart.name}"].acquire_lock.emit()
+        # only restart where necessary to avoid slowdowns
+        if (self.config_changed() or self.tls.upgrading) and self.cluster.started:
+            self.on[f"{self.restart.name}"].acquire_lock.emit()
```

Comment:
**praise** This is actually quite good. I kinda had the feeling that zookeeper was actually restarting too often, but it was more of a guess that actual need